### PR TITLE
Cr 118 receive case events from rm

### PR DESCRIPTION
**Pull request by cookel2**
# Motivation and Context

Receives CaseCreated and CaseUpdated type messages and deserialises them into relevant Java objects (CaseEvent objects) and then stores them, as JSON, in a bucket, named case_bucket, within the Google Cloud Platform. Also receives UACUpdated type messages and deserialises them into UACEvent objects and then stores them, as JSON, in a bucket (named uac_bucket) within the Google Cloud Platform.

# What has changed

Various Java classes have been added to the package named uk.gov.ons.ctp.integration.rhsvc.message.impl

JUnit tests have also been added or amended as appropriate (GCSDataStoreUnitTest, CaseEventReceiverImplTest, and UacEventReceiverImplTest)

# How to test?

Use the readme file to firstly set up Google credentials locally. Then run Rabbitmq locally using the command 'docker-compose up -d'. Use the following JSON format to put a CaseUpdated type message on the case-outbound-exchange (and use event.case.lifecycle as the routing key when you publish the message). You can then run the code and should see the message appear in Google Cloud Platform within a bucket named case_bucket NB. To see the buckets you need to select Storage --> Browser (within GCP). The code will create the bucket if it does not already exist (although there may be a problem if a bucket with the same name already exists in another project within GCP).

{
  "event" : {
    "type" : "CASE_UPDATED",
    "source" : "CASE_SERVICE",
    "channel" : "RM",
    "dateTime" : "2011-08-12T20:17:46.384Z",
    "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93"
  },
  "payload" : {
    "collectionCase" : {
        "id":"bbd55984-0dbf-4499-bfa7-0aa4228700e9",
        "caseRef":"10000000010",
        "survey":"CENSUS",
        "collectionExerciseId":"n66de4dc-3c3b-11e9-b210-d663bd873d93",
        "address": {
		"addressLine1":"1 main street",
  		"addressLine2":"upper upperingham",
  		"addressLine3":"",
  		"townName":"upton",
  		"postcode":"UP103UP",
  		"region":"E",    
  		"latitude":"50.863849",
  		"longitude":"-1.229710",
  		"uprn":"XXXXXXXXXXXXX",
  		"arid":"XXXXX",
  		"addressType":"CE",
  		"estabType":"XXX"
	},
	"contact": {
	"title":"Ms",
  	"forename":"jo",
  	"surname":"smith",
  	"email":"me@example.com",
  	"telNo":"+447890000000"
	},
        "state":"ACTIONABLE",
        "actionableFrom":"2011-08-12T20:17:46.384Z"
}
  }
}

Then use the following JSON format to put a UACUpdated type message on the case-outbound-exchange (and use event.uac.updates as the routing key when you publish the message). You can then run the code and should see the message appear in Google Cloud Platform within a bucket named uac_bucket. 

{
  "event" : {
    "type" : "UAC_UPDATED",
    "source" : "CASE_SERVICE",
    "channel" : "RM",
    "dateTime" : "2011-08-12T20:17:46.384Z",
    "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93"
  },
  "payload" : {
    "uac" : {
        "uacHash":  "72C84BA99D77EE766E9468A0DE36433A44888E5DEC4AFB84F8019777800B7364",        
        "active" : true,
        "questionnaireId":"1110000009",
        "caseType":"H",
        "region":"E",
        "caseId":"bbd55984-0dbf-4499-bfa7-0aa4228700e9",
        "collectionExerciseId":"n66de4dc-3c3b-11e9-b210-d663bd873d93"
}
  }
}

### CODE REVIEW

**Comment by cookel2** [comment]
I have finished making all the changes, and creating the JUnit tests, as requested by @philwhiles. Therefore this pull request is now ready for further review. See you tomorrow. Thanks, Ella

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceIT_Test.java [hunk 1]
```diff
@@ -6,87 +6,92 @@
...
+    collectionCase = new CollectionCase();
+    collectionCase.setId(caseId);
+    collectionCase.setCaseRef(caseRef);
+    collectionCase.setSurvey(survey);
+    collectionCase.setCollectionExerciseId(collectionExerciseId);
+    collectionCase.setSampleUnitRef(sampleUnitRef);
+    collectionCase.setAddress(addressLines);
+    collectionCase.setState(state);
+    collectionCase.setActionableFrom(actionableFrom);
+    //    collectionCase.setTimestamp(timestamp);
```
Comment (PeterBochel) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceIT_Test.java]: Dead code should be deleted. Also lines 36, 42 & 43

File: pom.xml [hunk 2]
```diff
@@ -196,9 +196,15 @@
...
-      <version>1.59.0</version>
+      <version>1.66.0</version>
     </dependency>
 
+    <!--<dependency>-->
+      <!--<groupId>com.google.cloud</groupId>-->
+      <!--<artifactId>google-cloud-storage</artifactId>-->
+      <!--<version>1.59.0</version>-->
+    <!--</dependency>-->
+
```
[the remark this answers is not in the document]
Comment (cookel2) [reply pom.xml]: commented entry now removed

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSDataStore.java [hunk 3]
```diff
@@ -28,15 +29,18 @@
...
     try {
+      log.with(bucket).info("Now attempting to create the bucket");
       createBucket(bucket, storage);
-
+      log.with(bucket).info("Bucket created successfully");
     } catch (StorageException se) {
-      // This Storage Exception is the only one declared on this API.
-      // If a bucket exists, this exception will be thrown
-      log.with(bucket).error("Bucket exists in the cloud storage.");
+      log.with(se.getMessage()).info("ERROR");
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSDataStore.java]: yes please raise a defect Peter, thanks

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSDataStore.java [hunk 4]
```diff
@@ -28,15 +29,18 @@
...
     }
+
+    log.info("Now saving the object to the cloud");
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSDataStore.java]: After discussion with Phil it has been agreed that entry and exit logging should be info level - other logging should be kept to a minimum and be debug level. I have now changed the "Now ssaving the object to the cloud" log message to include the bucket and the key.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/CloudDataStore.java [hunk 5]
```diff
@@ -5,8 +5,9 @@
 
 public interface CloudDataStore {
 
-  void storeObject(final String bucket, final String key, final String value)
-      throws StorageException;
+  void storeObject(final String bucket, final String key, final String value);
 
   Optional<String> retrieveObject(final String bucket, final String key) throws StorageException;
+
+  void deleteObject(final String key, final String bucket);
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/CloudDataStore.java]: now removed the 'throws StorageException' to make things consistent

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/CaseEvent.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class CaseEvent extends GenericEvent {
+
+  private CasePayload payload = new CasePayload();
+
+  @Override
+  public String toString() {
+    return super.toString() + " " + payload.toString();
+  }
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/CaseEvent.java]: removed the toString() method and replaced calls to it with explicit logging instead

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UACEventReceiverImpl.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+   * to prove messaging setup.
+   *
+   * @param uacEvent UACEvent message from Response Management
+   */
+  @ServiceActivator(inputChannel = "acceptUACEvent")
+  public void acceptUACEvent(UACEvent uacEvent) {
+
+    UAC uac;
+
+    log.info("The event being received is: " + uacEvent.toString());
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UACEventReceiverImpl.java]: This has now been replaced with explicit logging to just show the transactionId and event type (as discussed)

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UACEventReceiverImpl.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    uac = uacEvent.getPayload().getUac();
+
+    try {
+      respondentDataService.writeUAC(uac); // need to catch uk.gov.ons.ctp.common.error.CTPException
+    } catch (CTPException ctpEx) {
+      log.with(ctpEx.getMessage()).info("ERROR");
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UACEventReceiverImpl.java]: code now amended as suggested

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UacEventReceiverConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import org.mockito.Mockito;
+import org.springframework.amqp.rabbit.connection.Connection;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+import uk.gov.ons.ctp.integration.rhsvc.message.RespondentEventPublisher;
+
+@Configuration
+public class UacEventReceiverConfiguration {
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UacEventReceiverConfiguration.java]: There is only one UacEventReceiverConfiguration.java class that I can see. Do you mean that it is almost identical to the **Case**EventReceiverConfiguration class?

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceImpl.java [hunk 10]
```diff
@@ -35,45 +35,42 @@
...
    *
-   * @param uacContext - object to be serialised and stored in the cloud
+   * @param uac - object to be serialised and stored in the cloud
    * @throws CTPException - this is the exception thrown to the client.
    */
   @Override
-  public void writeUACContext(final UACContext uacContext) throws CTPException {
-    if (uacContext == null) {
+  public void writeUAC(final UAC uac) throws CTPException {
+    if (uac == null) {
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceImpl.java]: dead code now removed

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceIT_Test.java (hunk 1 above)
Comment (cookel2) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceIT_Test.java]: commented code now deleted

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/UACEvent.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class UACEvent extends GenericEvent {
+
+  private UACPayload payload = new UACPayload();
+
+  @Override
+  public String toString() {
+    return super.toString() + " " + payload.toString();
+  }
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/message/UACEvent.java]: toString() has now been removed as it's no longer needed for logging

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSDataStore.java [hunk 12]
```diff
@@ -70,6 +74,22 @@ public void storeObject(final String bucket, final String key, final String valu
...
+  @Override
+  public void deleteObject(final String key, final String bucket) {
+    log.info("Now in the deleteObject method in class GCSDataStore.");
+    if (null == bucket || bucket.length() == 0) {
+      log.with(bucket).info("Bucket name was not set for object retrieval");
+      return;
+    }
+    if (null == key || key.length() == 0) {
+      log.with(key).info("Key was not set for object retrieval");
+      return;
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSDataStore.java]: I have removed the null checks as these were not really necessary. I have added the key and bucket to the entry log to show what is being deleted and from where.

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UacEventReceiverConfiguration.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (cookel2) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/UacEventReceiverConfiguration.java]: I have now replaced the UacEventReceiverConfiguration and CaseEventReceiverConfiguration classes with a single class, named EventReceiverConfiguration.

**Comment by PeterBochel** [comment]
> Using strong typing highlights crap data we might be sent. If date time field sent is “Ella” would it be detected?

First the background and then a proposal. 

Ella and myself are both using the same Header class and it's my changes that have lead Ella to use a String for dateTime.

I've been focused on the outbound side which need to export values in the format 2011-08-12T20:17:46.384Z. Somebody has to know how to format a timestamp into the correct format. I believe that this can be done with custom Jackson formatting but that would embed the knowledge and dataTime formats into the DTO. (which breaks the convention of them being dumb transport objects).

The implications of storing the dateTime as a String are that getting the formatting done correctly is the either the responsibility of the code creating a Header or validation in the code receiving a header.

Perhaps we need a RHDateTime object? This would know how to convert to and from String representations. Would we need to write such a class, or does one of our open source projects already do this?

File: src/main/resources/springintegration/broker.xml [hunk 13]
```diff
@@ -40,15 +50,23 @@
...
 	<rabbit:direct-exchange name="case-deadletter-exchange">
 		<rabbit:bindings>
-			<rabbit:binding queue="Case.GatewayDLQ" key="Case.Gateway.binding" />
+			<rabbit:binding queue="Case.GatewayDLQ" key="event.case.lifecycle" />
+		</rabbit:bindings>
+	</rabbit:direct-exchange>
+
+	<rabbit:direct-exchange name="uac-deadletter-exchange">
+		<rabbit:bindings>
+			<rabbit:binding queue="UAC.GatewayDLQ" key="event.uac.updates" />
```
[the remark this answers is not in the document]
Comment (cookel2) [reply src/main/resources/springintegration/broker.xml]: Yes I agree. It is not something that I have looked at in detail - the existing code already used coldline storage. But it sounds sensible to make it configurable; not only the storage type but also the bucket name (as this needs to be unique across projects).

**Comment by cookel2** [comment]
> Not part of this commit, but I note the GCS bucket is hard coded to COLDLINE storage. This is as commented in the code the cheapest option but is specified as "Data you expect to access infrequently (i.e. no more than once per year)" and not suitable. Should perhaps be configurable and certainly default to Regional Storage.
> Also wonder whether access to GCS should be through a repository layer in the application rather than all be in the service layer. Something to discuss for future refactoring.

Yes I agree. It is not something that I have looked at in detail - the existing code already used coldline storage. But it sounds sensible to make it configurable; not only the storage type but also the bucket name (as this needs to be unique across projects).

**Comment by cookel2** [comment]
closed accidentally!

**Pull request merged**